File: 06-march-2024_Minimum_Remove_to_make_valid_Parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        stack<pair<char,int>> st;

        int n = s.length();

        for(int i=0;i<n;i++){
           if(s[i]=='('){
            st.push({'(',i}); 
           }
           if(s[i]==')'){
            if(st.empty()==true){
                s[i]='#';
            }else{
                st.pop();
            }
           }
        }

        while(st.empty()==false){
            s[st.top().second]='#';
            st.pop();
        }

        for(int i=0;i<n;i++){
          if(s[i]=='#')
          {
            s.erase(i,1);
            i--;
          }
        }

        return s;
    }
};

File: 06-march-2024_Minimum_Remove_to_make_valid_Parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        Stack<Pair> st = new Stack<>();

        int n = s.length();
        StringBuilder sb = new StringBuilder(s);

        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            if (ch == '(') {
                st.push(new Pair('(', i));
            }
            if (ch == ')') {
                if (st.isEmpty()) {
                    sb.setCharAt(i, '#');
                } else {
                    st.pop();
                }
            }
        }

        while (!st.isEmpty()) {
            sb.setCharAt(st.peek().index, '#');
            st.pop();
        }

        StringBuilder result = new StringBuilder();
        for (int i = 0; i < sb.length(); i++) {
            if (sb.charAt(i) != '#') {
                result.append(sb.charAt(i));
            }
        }

        return result.toString();
    }

    private class Pair {
        char bracket;
        int index;

        Pair(char bracket, int index) {
            this.bracket = bracket;
            this.index = index;
        }
    }
}
